File: packages/b2c-tooling-sdk/src/operations/code/ocapi-scripts-backend.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
import type {B2CInstance} from '../../instance/index.js';
import type {ScriptsBackend, CodeVersionInfo} from './scripts-types.js';
import type {CodeVersion as OcapiCodeVersion} from './versions.js';
import {
  listCodeVersions as ocapiListCodeVersions,
  getActiveCodeVersion as ocapiGetActiveCodeVersion,
  activateCodeVersion as ocapiActivateCodeVersion,
  deleteCodeVersion as ocapiDeleteCodeVersion,
  createCodeVersion as ocapiCreateCodeVersion,
  reloadCodeVersion as ocapiReloadCodeVersion,
} from './versions.js';

function mapOcapiCodeVersion(ocapi: OcapiCodeVersion): CodeVersionInfo {
  return {
    id: ocapi.id ?? '',
    active: ocapi.active,
    cartridges: ocapi.cartridges,
    compatibilityMode: ocapi.compatibility_mode,
    activationTime: ocapi.activation_time,
    lastModificationTime: ocapi.last_modification_time,
    rollback: ocapi.rollback,
    totalSize: ocapi.total_size,
    webDavUrl: ocapi.web_dav_url,
    _raw: ocapi,
  };
}

export class OcapiScriptsBackend implements ScriptsBackend {
  readonly name = 'ocapi' as const;

  constructor(private instance: B2CInstance) {}

  async listCodeVersions(): Promise<CodeVersionInfo[]> {
    const versions = await ocapiListCodeVersions(this.instance);
    return versions.map(mapOcapiCodeVersion);
  }

  async getActiveCodeVersion(): Promise<CodeVersionInfo | undefined> {
    const active = await ocapiGetActiveCodeVersion(this.instance);
    return active ? mapOcapiCodeVersion(active) : undefined;
  }

  async activateCodeVersion(codeVersionId: string): Promise<void> {
    await ocapiActivateCodeVersion(this.instance, codeVersionId);
  }

  async deleteCodeVersion(codeVersionId: string): Promise<void> {
    await ocapiDeleteCodeVersion(this.instance, codeVersionId);
  }

  async createCodeVersion(codeVersionId: string): Promise<void> {
    await ocapiCreateCodeVersion(this.instance, codeVersionId);
  }

  async reloadCodeVersion(codeVersionId?: string): Promise<void> {
    await ocapiReloadCodeVersion(this.instance, codeVersionId);
  }
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/operations/code/ocapi_scripts_backend.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""OCAPI implementation of the scripts (code-version) backend.

Mirrors ``src/operations/code/ocapi-scripts-backend.ts``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Literal

from b2c_tooling_sdk.clients.models.ocapi import CodeVersion as OcapiCodeVersion
from b2c_tooling_sdk.operations.code.scripts_types import CodeVersionInfo
from b2c_tooling_sdk.operations.code.versions import (
    CodeVersionActivationResult,
    activate_code_version,
    create_code_version,
    delete_code_version,
    get_active_code_version,
    list_code_versions,
)

if TYPE_CHECKING:
    from b2c_tooling_sdk.instance import B2CInstance


def _map_ocapi_code_version(version: OcapiCodeVersion) -> CodeVersionInfo:
    return CodeVersionInfo(
        id=version.id or "",
        active=version.active,
        cartridges=list(version.cartridges) if version.cartridges else None,
        compatibility_mode=version.compatibility_mode,
        activation_time=version.activation_time.isoformat() if version.activation_time else None,
        last_modification_time=version.last_modification_time.isoformat() if version.last_modification_time else None,
        rollback=version.rollback,
        total_size=version.total_size,
        web_dav_url=version.web_dav_url,
        raw=version,
    )


class OcapiScriptsBackend:
    """Code-version operations against the legacy OCAPI Data API."""

    def __init__(self, instance: B2CInstance) -> None:
        self._instance = instance

    @property
    def name(self) -> Literal["ocapi"]:
        return "ocapi"

    async def list_code_versions(self) -> list[CodeVersionInfo]:
        versions = await list_code_versions(self._instance)
        return [_map_ocapi_code_version(v) for v in versions]

    async def get_active_code_version(self) -> CodeVersionInfo | None:
        active = await get_active_code_version(self._instance)
        return _map_ocapi_code_version(active) if active else None

    async def activate_code_version(self, code_version_id: str) -> CodeVersionActivationResult:
        return await activate_code_version(self._instance, code_version_id)

    async def delete_code_version(self, code_version_id: str) -> None:
        await delete_code_version(self._instance, code_version_id)

    async def create_code_version(self, code_version_id: str) -> None:
        await create_code_version(self._instance, code_version_id)


__all__ = ["OcapiScriptsBackend"]
